src/components/page/page.component.tsx

```diff
@@ -1,24 +1,82 @@
 import { type Variants, motion } from "framer-motion";
-import type { PropsWithChildren } from "react";
+import { useEffect, useRef, useState, type PropsWithChildren } from "react";
 import styled from "styled-components";
+import { postEmbeddedMessage } from "~utils/embedded/utils/messages/embedded-messages.utils";
+import {
+  locationToRouteType,
+  routeTypeToPreferredLayout
+} from "~utils/embedded/utils/routes/embedded-routes.utils";
+import type { ArConnectRoutePath } from "~wallets/router/router.types";
 
 export interface PageProps extends PropsWithChildren {}
 
 export function Page({ children }: PageProps) {
+  const mainRef = useRef<HTMLDivElement>(null);
+
   const opacityAnimation: Variants = {
     initial: { opacity: 0 },
     enter: { opacity: 1 },
     exit: { opacity: 0, y: 0, transition: { duration: 0.2 } }
   };
 
+  const [height, setHeight] = useState(0);
+
+  useEffect(() => {
+    const mainElement = mainRef.current;
+
+    if (!mainElement || import.meta.env?.VITE_IS_EMBEDDED_APP !== "1") return;
+
+    const resizeObserver = new ResizeObserver((entries) => {
+      const size = entries?.[0]?.contentBoxSize?.[0] || {
+        inlineSize: 0,
+        blockSize: 0
+      };
+      const width = Math.ceil(size.inlineSize);
+      const height = Math.ceil(size.blockSize);
+
+      if (width > 0 && height > 0) {
+        // We get the path manually to avoid causing duplicate re-renders of the `Page` component if using the
+        // `useLocation` hook:
+        const path = location.hash.replace("#", "") as ArConnectRoutePath;
+        const routeType = locationToRouteType(path);
+        const preferredLayout = routeTypeToPreferredLayout(routeType);
+
+        postEmbeddedMessage({
+          type: "embedded_resize",
+          data: {
+            routeType,
+            preferredLayout,
+            width,
+            height
+          }
+        });
+
+        // For debugging only:;
+        setHeight(height);
+      }
+    });
+
+    resizeObserver.observe(mainElement);
```

Comment:
This will be notifying the parent about any change in size due to images loading, elements appearing/disappearing, etc.